# Data sources

| ID            | Name                                                              | Website                                                  | State            |
| ------------- | ----------------------------------------------------------------- | -------------------------------------------------------- | ---------------- |
| ABDAMED2      | ABDATA ABDAMED Drug Database                                      | [Link](https://abdata.de/datenangebot/abdamed/)          | Work in progress |
| CancerDrugsDB | A curated listing of licensed cancer drugs by the Anticancer Fund | [Link](https://www.anticancerfund.org/en/cancerdrugs-db) | Usable           |
| DGIdb         | The Drug Gene Interaction Database                                | [Link](http://dgidb.org)                                 | Work in progress |
| DisGeNET      | DisGeNET                                                          | [Link](https://www.disgenet.org)                         | Work in progress |
| DrugBank      | DrugBank                                                          | [Link](http://drugbank.ca)                               | Work in progress |
| DrugCentral   | DrugCentral                                                       | [Link](http://drugcentral.org)                           | Work in progress |
| GeneOntology  | Gene Ontology                                                     | [Link](http://geneontology.org)                          | Work in progress |
| HGNC          | HUGO Gene Nomenclature Committee                                  | [Link](https://www.genenames.org)                        | Usable           |
| HPO           | Human Phenotype Ontology                                          | [Link](https://hpo.jax.org)                              | Work in progress |
| ITIS          | Integrated Taxonomic Information System                           | [Link](https://www.itis.gov)                             | Work in progress |
| KEGG          | Kyoto Encyclopedia of Genes and Genomes                           | [Link](https://www.kegg.jp)                              | Work in progress |
| MED-RT        | Medication Reference Terminology (MED-RT™)                        | [Link](https://evs.nci.nih.gov/ftp1/MED-RT/)             | Usable           |
| Mondo         | Mondo Disease Ontology                                            | [Link](https://mondo.monarchinitiative.org)              | Work in progress |
| NDF-RT        | National Drug File – Reference Terminology (NDF-RT™)              | [Link](https://evs.nci.nih.gov/ftp1/NDF-RT/)             | Usable           |
| PharmGKB      | PharmGKB                                                          | [Link](https://www.pharmgkb.org)                         | Usable           |
| SIDER         | SIDER: Side Effect Resource                                       | [Link](http://sideeffects.embl.de)                       | Work in progress |
| UNII          | FDA Substance Registration System - Unique Ingredient Identifier  | [Link](https://fdasis.nlm.nih.gov/srs/)                  | Usable           |
| USDA-PLANTS   | USDA PLANTS Database                                              | [Link](https://plants.sc.egov.usda.gov)                  | Usable           |
|               |                                                                   |                                                          |                  |

## Licenses and references

### ABDAMED2

> ABDAMED is a licensed database and not freely available. Only license holders in posession of the database are able to use this integration module!

### CancerDrugsDB

> Pan Pantziarka, Rica Capistrano I, Arno De Potter, Liese Vandeborne, Gauthier Bouche (2021). An Open Access Database of Licensed Cancer Drugs Frontiers in Pharmacology, 11 March 2021 DOI: 10.3389/fphar.2021.627574

### DGIdb

> The source code for DGIdb is open-source and made available under the MIT license. The license is distributed with the source code (DGIdb license). The data used in DGIdb is all open access and where possible made available as raw data dumps in the downloads section.

> DGIdb 3.0: a redesign and expansion of the drug–gene interaction database. Cotto KC, Wagner AH, Feng Y, Kiwala S, Coffman AC, Spies G, Wollam A, Spies NC, Griffith OL, Griffith M. Nucleic Acids Research. 2017 Nov 17. doi: [10.1093/nar/gkx1143](https://doi.org/10.1093/nar/gkx1143).

### DisGeNET

> https://www.disgenet.org/legal

> The DisGeNET database is made available under the Attribution-NonCommercial-ShareAlike 4.0 International License whose text can be found here. If DisGeNET is incorporated into other works, we ask that DisGeNET is properly cited (see the citation guidelines ), and that the version number of DisGeNET is clearly displayed.

### DrugBank

> Creative Common’s Attribution-NonCommercial 4.0 International License

> Wishart DS, Knox C, Guo AC, Shrivastava S, Hassanali M, Stothard P, Chang Z, Woolsey J. Drugbank: a comprehensive resource for in silico drug discovery and exploration. Nucleic Acids Res. 2006 Jan 1;34 (Database issue):D668-72. 16381955.

### DrugCentral

> Creative Commons Attribution-ShareAlike 4.0 International Public License

> Ursu O, Holmes J, Knockel J, et al. DrugCentral: online drug compendium. Nucleic Acids Res. 2017;45(D1):D932-D939. doi: [10.1093/nar/gkw993](http://dx.doi.org/10.1093/nar/gkw993)

### GeneOntology

> Gene Ontology Consortium data and data products are licensed under the [Creative Commons Attribution 4.0 Unported License](https://creativecommons.org/licenses/by/4.0/legalcode)

> For citation information see http://geneontology.org/docs/go-citation-policy/

### HGNC

> https://www.genenames.org/about/

> **Data release policy**: No restrictions are imposed on access to, or use of, the data provided by the HGNC, which are provided to enhance knowledge and encourage progress in the scientific community. The HGNC provide these data in good faith, but make no warranty, express or implied, nor assume any legal liability or responsibility for any purpose for which they are used.

> **Guidelines on use of data in publications (copyright and licensing)**: It is a condition of our funding from NIH and the Wellcome Trust that the nomenclature and information we provide is freely available to all. Anyone may use the HGNC data, but we request that they reference the "HUGO Gene Nomenclature Committee at the European Bioinformatics Institute" and the website where possible.

### HPO

> https://hpo.jax.org/app/license

### ITIS

> Information presented on the ITIS website is considered public information and may be distributed or copied. Use of the appropriate citation is requested.

> To direct readers to the entire ITIS website, it is sufficient to give the address of the site in text:

> Retrieved [month, day, year], from the Integrated Taxonomic Information System (ITIS), www.itis.gov. https://doi.org/10.5066/F7KH0KBK

> To cite data obtained from ITIS, the following citation format is offered as a suggestion:

> Retrieved [month, day, year], from the Integrated Taxonomic Information System (ITIS) on-line database, www.itis.gov. https://doi.org/10.5066/F7KH0KBK

> Citations can be formatted to the style needed for your publication as long as there is sufficient information included to return the reader to the referenced information.

### KEGG

> A subset of KEGG MEDICUS is freely available at the GenomeNet FTP site. ftp://ftp.genome.jp/pub/kegg/medicus/

### MED-RT

> https://evs.nci.nih.gov/ftp1/MED-RT/medrt_about.html

> MED-RT has no copyright acknowledgement required. However, users must adhere to the UMLS license.

### Mondo

> https://mondo.monarchinitiative.org

> Mondo Disease Ontology is licensed under the Creative Commons Attribution 3.0 Unported (CC BY 3.0)

### NDF-RT

> https://www.nlm.nih.gov/research/umls/sourcereleasedocs/current/NDFRT/index.html

> NDF-RT is made available via the UMLS. Users of all UMLS ontologies must abide by the terms of the UMLS license, available at https://uts.nlm.nih.gov/license.html

### PharmGKB

> https://www.pharmgkb.org/page/dataUsagePolicy

> https://www.pharmgkb.org/page/citingPharmgkb

### SIDER

> Except as otherwise noted, SIDER is licensed under a Creative Commons Attribution-Noncommercial-Share Alike 4.0 License. For commercial use or customized versions, please contact biobyte solutions GmbH.

> Kuhn M, Letunic I, Jensen LJ, Bork P. The SIDER database of drugs and side effects. Nucleic Acids Res. 2015 Oct 19. doi: [10.1093/nar/gkv1075](http://dx.doi.org/10.1093/nar/gkv1075)

> Kuhn M, Campillos M, Letunic I, Jensen LJ, Bork P. A side effect resource to capture phenotypic effects of drugs. Mol Syst Biol. 2010;6:343. Epub 2010 Jan 19.

### UNII

> https://fdasis.nlm.nih.gov/srs/

> Except as otherwise noted, data is provided as Public Doman.

### USDA-PLANTS

> Plant information, including the distribution maps, lists, and text, is not copyrighted and is free for any use. Please cite the PLANTS Database as: USDA, NRCS. 2020. The PLANTS Database (http://plants.usda.gov, 23 November 2020). National Plant Data Team, Greensboro, NC 27401-4901 USA.
